vscode-dotty/src/tracer.ts

```diff
@@ -0,0 +1,370 @@
+        if (fs.existsSync(outputPath)) fs.unlinkSync(outputPath)
+        const output = fs.createWriteStream(outputPath)
+
+        const zip = archiver('zip')
+        zip.on('error', (err) => this.logError('zip error', safeError(err)))
+        zip.on('warning', (err) => this.logError('zip warning', safeError(err)))
+        zip.on('finish', () => {
+            this.ctx.extensionOut.appendLine('zip - finished')
+            fs.createReadStream(outputPath).pipe(
+                request.put(url, {
+                    qs: {
+                        client: this.machineId,
+                        project: this.projectId,
+                        session: this.sessionId
+                    }
+                })
+                    .on('error', (err) => this.logError('zip upload connection error', url, safeError(err)))
+                    .on('complete', (resp) => {
+                        if (!(resp.statusCode >= 200 && resp.statusCode < 300)) {
+                            this.logError('zip upload http error', url, resp.statusCode, resp.body)
+                        } else {
+                            this.ctx.extensionOut.appendLine('zip - http upload finished')
+                        }
+                    })
+            )
+        })
+
+        this.ctx.extensionOut.appendLine('zip - starting')
+        zip.pipe(output)
+        zip.glob('./**/*.{scala,sc,sbt,java}', { cwd: rootPath })
+        zip.glob('./**/.dotty-ide{.json,-artifact}', { cwd: rootPath })
+        zip.finalize()
+    }
+
+    private createRemoteLspOutputChannel(
+        remoteTracingUrl: string,
+        localOutputChannel: vscode.OutputChannel
+    ): vscode.OutputChannel {
+        const createSocket = () => {
+            const socket = new WebSocket(remoteTracingUrl, {
+                headers: {
+                    'X-DLS-Project-ID': this.projectId,
+                    'X-DLS-Client-ID': this.machineId,
+                    'X-DLS-Session-ID': this.sessionId,
+                },
+            })
+
+            const timer = setInterval(
+                () => {
+                    if (socket.readyState === WebSocket.OPEN) {
+                        socket.send('')
+                    } else if (socket.readyState === WebSocket.CLOSED) {
+                        clearInterval(timer)
+                    }
+                },
+                10 * 1000 /*ms*/,
+            )
+
+            socket.onerror = (event) => {
+                this.logErrorWithoutNotifying(
+                    'socket error',
+                    remoteTracingUrl,
+                    new SafeJsonifier(event, (event) => ({
+                        error: safeError(event.error),
+                        message: event.message,
+                        type: event.type
+                    }))
+                )
+                vscode.window.showWarningMessage('An error occured in Dotty LSP remote tracing connection.')
```

Comment:
I would rather not show any messages to users if something goes wrong, it's distracting and not very useful for them. What we could do is change the status bar message to have On/Off/Failed